brainframe_qt/api_utils/streaming/stream_manager.py

```diff
@@ -1,139 +1,175 @@
-            self._async_closing_streams.remove(stream)
-
-    def close_stream_async(self, stream_id: int) -> SyncedStreamReader:
-        stream = self._stream_readers.pop(stream_id)
-        self._async_closing_streams.append(stream)
-        stream.close()
-        return stream
-
-
-class StreamManagerAPI(BrainFrameAPI):
-    """Augments the API class to manage and provide a StreamManager."""
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self._stream_manager = None
-
-    def delete_stream_configuration(self, stream_id,
-                                    timeout=120):
-        super().delete_stream_configuration(stream_id, timeout)
-        if self._stream_manager is not None \
-                and self._stream_manager.is_streaming(stream_id):
-            self._stream_manager.close_stream_async(stream_id)
-
-    def get_stream_manager(self):
-        """Returns a singleton StreamManager object"""
-        # Lazily import streaming code to avoid OpenCV dependencies unless
-        # necessary
-        from brainframe_qt.api_utils.streaming import StreamManager
-
-        if self._stream_manager is None:
-            self._stream_manager = StreamManager(self.get_status_receiver())
-        return self._stream_manager
-
-    def get_stream_reader(self, stream_config: StreamConfiguration):
-        """Get the SyncedStreamReader for the given stream_configuration.
-
-        :param stream_config: The stream configuration to open.
-        :return: A SyncedStreamReader object
+        """Request and wait for all streams to close"""
+        logging.info("Initiating StreamManager close")
+        self._close()
+        logging.info("StreamManager close finished")
+
+    def delete_stream(self, stream_id: int, timeout: int = 120) -> None:
+        """[blocking API] Delete a stream through the API and initiate the closing of
+        its corresponding StreamReader
+        """
+        api.delete_stream_configuration(stream_id, timeout=timeout)
+        self.stop_streaming(stream_id)
+
+    def pause_streaming(self, stream_id) -> None:
+        self._set_stream_paused(stream_id, True)
+        self._ensure_running_streams()
+
+    def resume_streaming(self, stream_id) -> None:
+        self._set_stream_paused(stream_id, False)
+        self._ensure_running_streams()
+
+    def start_streaming(
+            self,
+            stream_conf: StreamConfiguration,
+            url: str,
+    ) -> SyncedStreamReader:
+        """Starts reading from the stream using the given information, or returns an
+        existing reader if we're already reading this stream.
+
+        :param stream_conf: The stream to connect to
+        :param url: The URL to stream on
+        :return: A SyncedStreamReader for the stream
         """
-        url = self.get_stream_url(stream_config.id)
-        logging.info("API: Opening stream on url " + url)
+        return self._start_stream(stream_conf, url)
```

Comment:
Can't `start_streaming` and `_start_streaming` be combined as well?

Reply:
Same as `close` comment